proposals/wg-image-compatibility.md

```diff
@@ -0,0 +1,106 @@
+Accordingly, the WG will operate in accordance with the OCI Charter and OCI's other policies and procedures, supplemented by the details below.
+
+## Purpose
+
+In mission-critical industries, applications often require special features provided by host operating systems.
+These are usually described in the release notes or installation guides. We believe we have identified the missing puzzle in the OCI standard,
+the image compatibility specification, which is a key feature that describes the requirements of special containerized applications
+that have hard requirements for specific kernel versions, configurations, modules, or out-of-tree drivers.
+
+For example, a container application  free5GC UPF requiring the gtp5g kernel module  will only be compatible with kernel 5.0.0-23-generic 
+or 5.4.x due to the module's hard kernel version requirements - https://free5gc.org/guide/3-install-free5gc/.
+If SR-IOV or GPU is required by the container, other requirements have to be specified against the host OS.
+Kata Containers use cases are the best example: https://github.com/kata-containers/kata-containers/tree/main/docs/use-cases.
+
+
+The described incompatibility problem affects distributed systems.
+In multi-cloud, hybrid cloud, or even private cloud scenarios, one containerized cloud-native application will need to run
+on different Kubernetes distributions that use different host operating systems on different worker nodes.
+
+Operating systems can use different kernel versions with different configurations.
+Different operating system distributions may integrate different devices and their corresponding drivers,
+even for the same device different versions of drivers may be used. Therefore, different operating system distributions are not identical.
+Even within the same OS distro, different releases are also not identical.
+
+An image compatibility specification would help container image authors describe compatibility requirements in a standard way.
+The specification will be uploaded with the image to the image registry. This makes hard container compatibility requirements discoverable,
+programmable, and will support different consumers and cover use cases where the application requires a specific compatible environment.
+
+For further reference please follow:
+* [OCI Image Compatibility - the spec reasoning](https://docs.google.com/presentation/d/1F9GnCm2sULuyTJ5BEFZlL8Qjab81DK7g9Oy6qBfe5Qs/edit?usp=sharing)
+* [Revisit container and host OS compatibility](https://docs.google.com/document/d/1lzwh8DGMu5vXXHwJmnewYIMffkcOEvH8owX4UYjRcw0/edit#heading=h.fcxt9vheg92c)
+
+## Scope
+
+* Define image compatibility spec that describes special host OS requirements of containerized application.
+* The spec should describe container requirements for Linux, FreeBSD and Windows.
+* The final list of supported fields (like kernel modules, configuration, out-of-tree drivers etc.)
```

Comment:
Love it! Over in HPC land we are also concerned about optimised binaries. We thought of putting the optimisation description in annotations so that the runtime can check (and fail if the image is optimised for a non-compliant micro-arch) and a registry can pick the right image. [Blog post about that](https://blog.qnib.org/blog/2023/01/30/the-hpc-container-conformance-project/)

Reply:
at first, I also have thought about using annotations/labels to describe the compatibility requirements. But I found in some scenario annotation way is not flexible: the annotation is built into the image by the author, if the image is compatible with new release kernel or device drivers, and the annotations need to updated, then the author has to release a new version of image. If artifact is used instead, then the image can keep unchanged, just replace the artifact(or attach a new version artifact). This is quite useful in production environment, the production environment can keep application running while introducing new OS distribution, and no need to test the new image. The another one is artifact can provide possibibility to describe application(a set of images) level compatibility, annotations will work on single image granularity, this topic needs more discussion after working group is created.

Reply:
@ChristianKniep would you like to join the wg?

Reply:
Indeed I would :)

Reply:
@mfranczy I am interested too, but I need to get a sense of the meeting time and frequency.

Reply:
@ChaoyiHuang What I like about the annotations is that everyone can have their own... :) 
Back in the days I used the `features` field in platform to encode information about the micro-arch. https://github.com/moby/moby/issues/38715
The purpose of the `features` field might have been to specific. AFAIK it was envisioned to encapsulate the CPU flags at one point.
I'd like for us to not fall in the same trap. Hammering out a spec that covers all use-cases might be to ambitious.

Every domain/field can go as shallow and deep into specific annotations as they like. Our idea with annotations was that we can bake them into OCI Manifests AND Indexes. 
We can even carry them as LABELS within Docker images. They can be used to smoke-test a potential image candiate to see if the node matches all requirements and expectations. 
And even before. I need to pick a manifest from a given index with manifest of the same platform but with different annotations.

Reply:
@vsoch once a week should be enough. Time I don't know yet, first TOB will have to agree to form this working group, then we will have to find a common ground about time. We will have multiple time zones so I think to create a poll to see what works.

@vsoch, @ChristianKniep let me add you to the owners. What about stakeholders, what should I put in there?

Reply:
Stakeholders are our employers? If so I’d need to ask first.

Reply:
> Stakeholders are our employers? If so I’d need to ask first.

https://github.com/opencontainers/tob/pull/128/files/7e2ea045dbbe650c66a249124bf7750ee7374fec#r1324498485

As suggested in the comment, list of likely adopters and implementers of the change. If you need the compatibility spec to influence any specific product then we can add it. Additionally, if you have agreement with your employer to work on this project then I think it would be fine to put them too. If you do it on your own we don't have to place them.

Reply:
> Every domain/field can go as shallow and deep into specific annotations as they like. Our idea with annotations was that we can bake them into OCI Manifests AND Indexes.
> We can even carry them as LABELS within Docker images. They can be used to smoke-test a potential image candiate to see if the node matches all requirements and expectations.
> And even before. I need to pick a manifest from a given index with manifest of the same platform but with different annotations.

how to define spec is a little too far, the first step is to create a working group :)

Reply:
Individuals that want to participate would be listed under "Proposed Owners". "Stakeholders" are the projects that would implement it.